### ISSUE TITLE
Hyprland now crashes on removing and reconnecting monitor

### ISSUE THREAD

**Issue by justinesmithies** [hyprwm/Hyprland#1806]
I used to be able to unplug my external monitor and plug it back in with no issues. But recent updates make Hyprland crash when I plug back in the monitor after unplugging it.

```
--------------------------------------------
   Hyprland Crash Report
--------------------------------------------
Well this is awkward.

Hyprland received signal 11 (Segmentation fault)

System info:
	System name: Linux
	Node name: justine-laptop
	Release: 6.2.6-arch1-1
	Version: #1 SMP PREEMPT_DYNAMIC Mon, 13 Mar 2023 17:02:08 +0000

GPU:
	07:00.0 VGA compatible controller [0300]: Advanced Micro Devices, Inc. [AMD/ATI] Renoir [1002:1636] (rev d1) (prog-if 00 [VGA controller])


os-release:
	NAME="Arch Linux"
	PRETTY_NAME="Arch Linux"
	ID=arch
	BUILD_ID=rolling
	ANSI_COLOR="38;2;23;147;209"
	HOME_URL="https://archlinux.org/"
	DOCUMENTATION_URL="https://wiki.archlinux.org/"
	SUPPORT_URL="https://bbs.archlinux.org/"
	BUG_REPORT_URL="https://bugs.archlinux.org/"
	PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
	LOGO=archlinux-logo
	


Backtrace:
	#0 | Hyprland(_ZN13CrashReporter18createAndSaveCrashEi+0x687) [0x555c2b85bd77]
		??
		??:0
	#1 | Hyprland(_Z25handleUnrecoverableSignali+0x3c) [0x555c2b7fcc1c]
		??
		??:0
	#2 | /usr/lib/libc.so.6(+0x38f50) [0x7fef97569f50]
		??
		??:0
	#3 | /usr/lib/libwlroots.so.12032(wlr_output_damage_add+0xe) [0x7fef97e4ff5e]
		??
		??:0
	#4 | /usr/lib/libwlroots.so.12032(wlr_output_damage_add_box+0x3d) [0x7fef97e5017d]
		??
		??:0
	#5 | Hyprland(_ZN13CHyprRenderer9damageBoxEP7wlr_box+0xc5) [0x555c2b903065]
		??
		??:0
	#6 | Hyprland(_ZN25CHyprDropShadowDecoration12damageEntireEv+0x157) [0x555c2b90a357]
		??
		??:0
	#7 | Hyprland(_ZN25CHyprDropShadowDecoration12updateWindowEP7CWindow+0x24) [0x555c2b909c14]
		??
		??:0
	#8 | Hyprland(_ZN11CCompositor36updateWindowAnimatedDecorationValuesEP7CWindow+0x4cf) [0x555c2b803bdf]
		??
		??:0
	#9 | Hyprland(_ZN11CCompositor11focusWindowEP7CWindowP11wlr_surface+0x17e) [0x555c2b80697e]
		??
		??:0
	#10 | Hyprland(_ZN13CInputManager16mouseMoveUnifiedEjb+0xe6d) [0x555c2b8d304d]
		??
		??:0
	#11 | Hyprland(_ZN13CHyprRenderer16applyMonitorRuleEP8CMonitorP12SMonitorRuleb+0x54b) [0x555c2b907a3b]
		??
		??:0
	#12 | Hyprland(_ZN8CMonitor9onConnectEb+0x5e0) [0x555c2b88fd60]
		??
		??:0
	#13 | Hyprland(_ZN6Events18listener_newOutputEP11wl_listenerPv+0x107) [0x555c2b877107]
		??
		??:0
	#14 | /usr/lib/libwayland-server.so.0(wl_signal_emit_mutable+0x7e) [0x7fef97cc4f6e]
		??
		??:0
	#15 | /usr/lib/libwayland-server.so.0(wl_signal_emit_mutable+0x7e) [0x7fef97cc4f6e]
		??
		??:0
	#16 | /usr/lib/libwlroots.so.12032(+0x54919) [0x7fef97e11919]
		??
		??:0
	#17 | /usr/lib/libwayland-server.so.0(wl_signal_emit_mutable+0x7e) [0x7fef97cc4f6e]
		??
		??:0
	#18 | /usr/lib/libwlroots.so.12032(+0x4e870) [0x7fef97e0b870]
		??
		??:0
	#19 | /usr/lib/libwayland-server.so.0(wl_event_loop_dispatch+0xa2) [0x7fef97cc69e2]
		??
		??:0
	#20 | /usr/lib/libwayland-server.so.0(wl_display_run+0x27) [0x7fef97cc7197]
		??
		??:0
	#21 | Hyprland(main+0x5ad) [0x555c2b7f062d]
		??
		??:0
	#22 | /usr/lib/libc.so.6(+0x23790) [0x7fef97554790]
		??
		??:0
	#23 | /usr/lib/libc.so.6(__libc_start_main+0x8a) [0x7fef9755484a]
		??
		??:0
	#24 | Hyprland(_start+0x25) [0x555c2b7fcae5]
		??
		??:0


Log tail:
[LOG] Changed to workspace 3
[LOG] moveWorkspaceToMonitor: SWITCHINGISACTIVE, active 3 -> 2
[LOG] Removing monitor HDMI-A-1 from realMonitors
[LOG] Callback 2e210bc8 -> 2e210bc0,  removed.
[LOG] Callback 2e210b60 -> 2e210b58,  removed.
[LOG] Destroying workspace ID 3
[LOG] Fading out a non-existent LS??
[LOG] LayerSurface 2dfc78b0 unmapped
[LOG] Framebuffer created, status 36053
[LOG] LayerSurface 2dfc78b0 destroyed
[LOG] Callback 2e2191b8 -> 2e2191b0, layerSurface removed.
[LOG] Callback 2e219080 -> 2e219078, layerSurface removed.
[LOG] Callback 2e2190e8 -> 2e2190e0, layerSurface removed.
[LOG] Callback 2e219150 -> 2e219148, layerSurface removed.
[LOG] Callback 2e219220 -> 2e219218, layerSurface removed.
[LOG] LayerSurface 2e293350 arranged: x: 1920 y: 0 w: 1920 h: 1080 with margins: t: 0 l: 0 r: 0 b: 0
[LOG] Monitor eDP-1 layers arranged: reserved: 0.000000 0.000000 0.000000 0.000000
[LOG] Fading out a non-existent LS??
[LOG] Hiding the cursor (timeout)
[LOG] Cleanup: destroyed a layersurface
[LOG] Registered signal for owner 2e219040: 2dfc7818 -> 2e219190 (owner: layerSurface)
[LOG] Registered signal for owner 2e219040: 2eabee50 -> 2e219058 (owner: layerSurface)
[LOG] Registered signal for owner 2e219040: 2eabee60 -> 2e2190c0 (owner: layerSurface)
[LOG] Registered signal for owner 2e219040: 2eabee70 -> 2e219128 (owner: layerSurface)
[LOG] Registered signal for owner 2e219040: 2eabee80 -> 2e2191f8 (owner: layerSurface)
[LOG] LayerSurface 2eabed90 (namespace gtk-layer-shell layer 1) created on monitor eDP-1
[LOG] LayerSurface 2e264e10 arranged: x: 1920 y: 10 w: 1920 h: 25 with margins: t: 10 l: 0 r: 0 b: 0
[LOG] LayerSurface 2e293350 arranged: x: 1920 y: 0 w: 1920 h: 1080 with margins: t: 0 l: 0 r: 0 b: 0
[LOG] Monitor eDP-1 layers arranged: reserved: 0.000000 35.000000 0.000000 0.000000
[LOG] LayerSurface 2eabed90 mapped
[LOG] LayerSurface 2e264e10 arranged: x: 1920 y: 10 w: 1920 h: 21 with margins: t: 10 l: 0 r: 0 b: 0
[LOG] LayerSurface 2e293350 arranged: x: 1920 y: 0 w: 1920 h: 1080 with margins: t: 0 l: 0 r: 0 b: 0
[LOG] Monitor eDP-1 layers arranged: reserved: 0.000000 31.000000 0.000000 0.000000
[LOG] LayerSurface 2e264e10 arranged: x: 1920 y: 10 w: 1920 h: 21 with margins: t: 10 l: 0 r: 0 b: 0
[LOG] LayerSurface 2e293350 arranged: x: 1920 y: 0 w: 1920 h: 1080 with margins: t: 0 l: 0 r: 0 b: 0
[LOG] Monitor eDP-1 layers arranged: reserved: 0.000000 31.000000 0.000000 0.000000
[LOG] Socket 2 accepted a new client at FD 109
[LOG] LayerSurface 2e264e10 arranged: x: 1920 y: 10 w: 1920 h: 29 with margins: t: 10 l: 0 r: 0 b: 0
[LOG] LayerSurface 2e293350 arranged: x: 1920 y: 0 w: 1920 h: 1080 with margins: t: 0 l: 0 r: 0 b: 0
[LOG] Monitor eDP-1 layers arranged: reserved: 0.000000 39.000000 0.000000 0.000000
[LOG] Adding completely new monitor.
[LOG] Registered signal for owner 2e210980: 2df8bda0 -> 2e210ad0 (owner: )
[LOG] Registered signal for owner 2e210980: 2df8be30 -> 2e210b38 (owner: )
[LOG] Registered signal for owner 2e210980: 2df8be20 -> 2e210ba0 (owner: )
[LOG] Applying monitor rule for HDMI-A-1
[LOG] Monitor HDMI-A-1: requested 1920x1080@60.000000, found available mode: 1920x1080@60000mHz, applying.
[LOG] Monitor HDMI-A-1 -> destroyed all render data
[LOG] Monitor HDMI-A-1 layers arranged: reserved: 0.000000 0.000000 0.000000 0.000000
[LOG] Monitor HDMI-A-1 data dump: res 1920x1080@60.00Hz, scale 1.00, transform 0, pos 0x0, 10b 0
[LOG] Set keyboard focus to surface 2e257470, with window name: justine@justine-laptop:~
```

**Comment by vaxerski** [comment]
can't repro. You sure you on latest git?

**Comment by justinesmithies** [comment]
I'm definitely on the latest git master and i open foot on one screen then qutebrowser on the other. I then umplug the external HDMI screen and everything is ok but when I reconnect the HDMI Hyprland crashes out. Log below.

```
--------------------------------------------
   Hyprland Crash Report
--------------------------------------------
I tried so hard, and got so far...

Hyprland received signal 11 (Segmentation fault)

Version: 5b924aaf608d91e411bd4c05751dc0b7323c36a7

System info:
	System name: Linux
	Node name: justine-laptop
	Release: 6.2.6-arch1-1
	Version: #1 SMP PREEMPT_DYNAMIC Mon, 13 Mar 2023 17:02:08 +0000

GPU:
	07:00.0 VGA compatible controller [0300]: Advanced Micro Devices, Inc. [AMD/ATI] Renoir [1002:1636] (rev d1) (prog-if 00 [VGA controller])


os-release:
	NAME="Arch Linux"
	PRETTY_NAME="Arch Linux"
	ID=arch
	BUILD_ID=rolling
	ANSI_COLOR="38;2;23;147;209"
	HOME_URL="https://archlinux.org/"
	DOCUMENTATION_URL="https://wiki.archlinux.org/"
	SUPPORT_URL="https://bbs.archlinux.org/"
	BUG_REPORT_URL="https://bugs.archlinux.org/"
	PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
	LOGO=archlinux-logo
	


Backtrace:
	#0 | Hyprland(_ZN13CrashReporter18createAndSaveCrashEi+0x721) [0x55ea1ed1fe31]
		??
		??:0
	#1 | Hyprland(_Z25handleUnrecoverableSignali+0x3c) [0x55ea1ecc0c3c]
		??
		??:0
	#2 | /usr/lib/libc.so.6(+0x38f50) [0x7f97b0c51f50]
		??
		??:0
	#3 | /usr/lib/libwlroots.so.12032(wlr_output_damage_add+0xe) [0x7f97b1553f5e]
		??
		??:0
	#4 | /usr/lib/libwlroots.so.12032(wlr_output_damage_add_box+0x3d) [0x7f97b155417d]
		??
		??:0
	#5 | Hyprland(_ZN13CHyprRenderer9damageBoxEP7wlr_box+0xc5) [0x55ea1edc7115]
		??
		??:0
	#6 | Hyprland(_ZN25CHyprDropShadowDecoration12damageEntireEv+0x157) [0x55ea1edce407]
		??
		??:0
	#7 | Hyprland(_ZN25CHyprDropShadowDecoration12updateWindowEP7CWindow+0x24) [0x55ea1edcdcc4]
		??
		??:0
	#8 | Hyprland(_ZN11CCompositor36updateWindowAnimatedDecorationValuesEP7CWindow+0x4cf) [0x55ea1ecc7bff]
		??
		??:0
	#9 | Hyprland(_ZN11CCompositor11focusWindowEP7CWindowP11wlr_surface+0x17e) [0x55ea1ecca99e]
		??
		??:0
	#10 | Hyprland(_ZN13CInputManager16mouseMoveUnifiedEjb+0xe6d) [0x55ea1ed970fd]
		??
		??:0
	#11 | Hyprland(_ZN13CHyprRenderer16applyMonitorRuleEP8CMonitorP12SMonitorRuleb+0x54b) [0x55ea1edcbaeb]
		??
		??:0
	#12 | Hyprland(_ZN8CMonitor9onConnectEb+0x5e0) [0x55ea1ed53e10]
		??
		??:0
	#13 | Hyprland(_ZN6Events18listener_newOutputEP11wl_listenerPv+0x107) [0x55ea1ed3b1b7]
		??
		??:0
	#14 | /usr/lib/libwayland-server.so.0(wl_signal_emit_mutable+0x7e) [0x7f97b13c8f6e]
		??
		??:0
	#15 | /usr/lib/libwayland-server.so.0(wl_signal_emit_mutable+0x7e) [0x7f97b13c8f6e]
		??
		??:0
	#16 | /usr/lib/libwlroots.so.12032(+0x54919) [0x7f97b1515919]
		??
		??:0
	#17 | /usr/lib/libwayland-server.so.0(wl_signal_emit_mutable+0x7e) [0x7f97b13c8f6e]
		??
		??:0
	#18 | /usr/lib/libwlroots.so.12032(+0x4e870) [0x7f97b150f870]
		??
		??:0
	#19 | /usr/lib/libwayland-server.so.0(wl_event_loop_dispatch+0xa2) [0x7f97b13ca9e2]
		??
		??:0
	#20 | /usr/lib/libwayland-server.so.0(wl_display_run+0x27) [0x7f97b13cb197]
		??
		??:0
	#21 | Hyprland(main+0x5ad) [0x55ea1ecb464d]
		??
		??:0
	#22 | /usr/lib/libc.so.6(+0x23790) [0x7f97b0c3c790]
		??
		??:0
	#23 | /usr/lib/libc.so.6(__libc_start_main+0x8a) [0x7f97b0c3c84a]
		??
		??:0
	#24 | Hyprland(_start+0x25) [0x55ea1ecc0b05]
		??
		??:0


Log tail:
[LOG] Changed to workspace 3
[LOG] moveWorkspaceToMonitor: SWITCHINGISACTIVE, active 3 -> 1
[LOG] Removing monitor HDMI-A-1 from realMonitors
[LOG] Callback 2035b948 -> 2035b940,  removed.
[LOG] Callback 2035b8e0 -> 2035b8d8,  removed.
[LOG] Destroying workspace ID 3
[LOG] Fading out a non-existent LS??
[LOG] LayerSurface 20ddb540 unmapped
[LOG] Framebuffer created, status 36053
[LOG] LayerSurface 20ddb540 destroyed
[LOG] Callback 205a2728 -> 205a2720, layerSurface removed.
[LOG] Callback 205a25f0 -> 205a25e8, layerSurface removed.
[LOG] Callback 205a2658 -> 205a2650, layerSurface removed.
[LOG] Callback 205a26c0 -> 205a26b8, layerSurface removed.
[LOG] Callback 205a2790 -> 205a2788, layerSurface removed.
[LOG] LayerSurface 205c4200 arranged: x: 1920 y: 0 w: 1920 h: 1080 with margins: t: 0 l: 0 r: 0 b: 0
[LOG] Monitor eDP-1 layers arranged: reserved: 0.000000 0.000000 0.000000 0.000000
[LOG] Fading out a non-existent LS??
[LOG] Cleanup: destroyed a layersurface
[LOG] Registered signal for owner 205a25b0: 2055aa08 -> 205a2700 (owner: layerSurface)
[LOG] Registered signal for owner 205a25b0: 20576bf0 -> 205a25c8 (owner: layerSurface)
[LOG] Registered signal for owner 205a25b0: 20576c00 -> 205a2630 (owner: layerSurface)
[LOG] Registered signal for owner 205a25b0: 20576c10 -> 205a2698 (owner: layerSurface)
[LOG] Registered signal for owner 205a25b0: 20576c20 -> 205a2768 (owner: layerSurface)
[LOG] LayerSurface 20576b30 (namespace gtk-layer-shell layer 1) created on monitor eDP-1
[LOG] LayerSurface 205c1180 arranged: x: 1920 y: 10 w: 1920 h: 25 with margins: t: 10 l: 0 r: 0 b: 0
[LOG] LayerSurface 205c4200 arranged: x: 1920 y: 0 w: 1920 h: 1080 with margins: t: 0 l: 0 r: 0 b: 0
[LOG] Monitor eDP-1 layers arranged: reserved: 0.000000 35.000000 0.000000 0.000000
[LOG] LayerSurface 20576b30 mapped
[LOG] LayerSurface 205c1180 arranged: x: 1920 y: 10 w: 1920 h: 21 with margins: t: 10 l: 0 r: 0 b: 0
[LOG] LayerSurface 205c4200 arranged: x: 1920 y: 0 w: 1920 h: 1080 with margins: t: 0 l: 0 r: 0 b: 0
[LOG] Monitor eDP-1 layers arranged: reserved: 0.000000 31.000000 0.000000 0.000000
[LOG] LayerSurface 205c1180 arranged: x: 1920 y: 10 w: 1920 h: 21 with margins: t: 10 l: 0 r: 0 b: 0
[LOG] LayerSurface 205c4200 arranged: x: 1920 y: 0 w: 1920 h: 1080 with margins: t: 0 l: 0 r: 0 b: 0
[LOG] Monitor eDP-1 layers arranged: reserved: 0.000000 31.000000 0.000000 0.000000
[LOG] Socket 2 accepted a new client at FD 113
[LOG] LayerSurface 205c1180 arranged: x: 1920 y: 10 w: 1920 h: 29 with margins: t: 10 l: 0 r: 0 b: 0
[LOG] LayerSurface 205c4200 arranged: x: 1920 y: 0 w: 1920 h: 1080 with margins: t: 0 l: 0 r: 0 b: 0
[LOG] Monitor eDP-1 layers arranged: reserved: 0.000000 39.000000 0.000000 0.000000
[LOG] Hiding the cursor (timeout)
[LOG] Adding completely new monitor.
[LOG] Registered signal for owner 2035b700: 202feb10 -> 2035b850 (owner: )
[LOG] Registered signal for owner 2035b700: 202feba0 -> 2035b8b8 (owner: )
[LOG] Registered signal for owner 2035b700: 202feb90 -> 2035b920 (owner: )
[LOG] Applying monitor rule for HDMI-A-1
[LOG] Monitor HDMI-A-1: requested 1920x1080@60.000000, found available mode: 1920x1080@60000mHz, applying.
[LOG] Monitor HDMI-A-1 -> destroyed all render data
[LOG] Monitor HDMI-A-1 layers arranged: reserved: 0.000000 0.000000 0.000000 0.000000
[LOG] Monitor HDMI-A-1 data dump: res 1920x1080@60.00Hz, scale 1.00, transform 0, pos 0x0, 10b 0
[LOG] Set keyboard focus to surface 20dc4940, with window name: hyprwm/Hyprland: Hyprland is a highly customizable dynamic tiling Wayland compositor that doesn't sacrifice on its looks. - qutebrowser
```

**Comment by vaxerski** [comment]
[patch.txt](https://github.com/hyprwm/Hyprland/files/11004051/patch.txt)
try this

**Comment by justinesmithies** [comment]
Hopefully I did right by applying that patch using the PKGBUILD ? But it seems to be working ok now I can't make it crash ATM ??? See what I did to the PKGBUILD to see if that was right ;)

```
# Maintainer: ThatOneCalculator <kainoa@t1c.dev>

_pkgname="hyprland"
pkgname="${_pkgname}-git"
pkgver=r2607.5b924aaf
pkgrel=1
pkgdesc="A dynamic tiling Wayland compositor based on wlroots that doesn't sacrifice on its looks."
arch=(any)
url="https://github.com/hyprwm/Hyprland"
license=('BSD')
depends=(
	libxcb
	xcb-proto
	xcb-util
	xcb-util-keysyms
	libxfixes
	libx11
	libxcomposite
	xorg-xinput
	libxrender
	pixman
	wayland-protocols
	cairo
	pango
	polkit
	glslang
	libinput
	libxcb
	libxkbcommon
	opengl-driver
	pixman
	wayland
	xcb-util-errors
	xcb-util-renderutil
	xcb-util-wm
	seatd
	vulkan-icd-loader
	vulkan-validation-layers
	xorg-xwayland
	libliftoff
	libdisplay-info)
makedepends=(
	git
	cmake
	ninja
	gcc
	gdb
	meson
	vulkan-headers
	wayland-protocols
	xorgproto)
source=("${_pkgname}::git+https://github.com/hyprwm/Hyprland.git")
conflicts=("${_pkgname}")
provides=(hyprland)
sha256sums=('SKIP')
options=(!makeflags !buildflags !strip)

pkgver() {
  cd "$_pkgname"
  ( set -o pipefail
    git describe --long 2>/dev/null | sed 's/\([^-]*-g\)/r\1/;s/-/./g' ||
    printf "r%s.%s" "$(git rev-list --count HEAD)" "$(git rev-parse --short HEAD)"
  )
}

prepare() {
    cd "${srcdir}/${_pkgname}"
    patch --forward --strip=1 --input="/home/justine/Downloads/patch.txt"
}

build() {
	cd "${srcdir}/${_pkgname}"
	git submodule update --init
	make fixwlr
	cd "./subprojects/wlroots/" && meson build/ --prefix="${srcdir}/tmpwlr" --buildtype=release && ninja -C build/ && mkdir -p "${srcdir}/tmpwlr" && ninja -C build/ install && cd ../
	cd udis86 && cmake --no-warn-unused-cli -DCMAKE_BUILD_TYPE:STRING=Release -H./ -B./build -G Ninja && cmake --build ./build --config Release --target all -j$(shell nproc) && cd ../..
	make protocols
    make release
	cd ./hyprctl && make all && cd ..
}

package() {
	cd "${srcdir}/${_pkgname}"
	mkdir -p "${pkgdir}/usr/share/wayland-sessions"
	mkdir -p "${pkgdir}/usr/share/hyprland"
	install -Dm755 build/Hyprland -t "${pkgdir}/usr/bin"
	install -Dm755 hyprctl/hyprctl -t "${pkgdir}/usr/bin"
	install -Dm644 assets/*.png -t "${pkgdir}/usr/share/hyprland"
	install -Dm644 example/hyprland.desktop -t "${pkgdir}/usr/share/wayland-sessions"
	install -Dm644 example/hyprland.conf -t "${pkgdir}/usr/share/hyprland"
	install -Dm644 LICENSE -t "${pkgdir}/usr/share/licenses/${_pkgname}"
	install -Dm755 ../tmpwlr/lib/libwlroots.so.12032 -t "${pkgdir}/usr/lib"
}
```

**Comment by justinesmithies** [comment]
I can no longer make it crash after that patch was applied.

**Comment by vaxerski** [comment]
fixed in 91a565c7b0957968d686fccf8f3a5166b9c3fec9 then
